### ISSUE TITLE
s3cmd put sometimes crashes when upload fails and it retries upload

### ISSUE THREAD

**Issue by bigwillystyle42** [s3tools/s3cmd#1074]
We upload logs from several hundred servers around the world. Most of them are great, no issues. Some of them have less than ideal internet connections (packet loss, small pipe). When s3cmd attempts to upload a file on our not as great servers, sometimes the upload times out. When s3cmd tries again, it frequently aborts or segfaults.

We're using s3cmd 2.0.2-1 from the Debian repositories on Debian 9 64bit servers. This is difficult to reproduce, but it happens frequently enough on some of our servers to test fixes or suggestions. I do not believe we had this issue on s3cmd 1.5.x, we were forced to upgrade because we started using a new region which only supports v4 signatures.

Here's example output from an Abort:
```
WARNING: Upload failed: /year%3D2020/month%3D02/day%3D22/server%3Dbangkok/2020-02-22-13-00-07-386017311-97cb2717-squid.log.gz (400 (RequestTimeout): Your socket connection to the server was not read from or written to within the timeout period. Idle connections will be closed.)
WARNING: Waiting 3 sec...
WARNING: Upload failed: /year%3D2020/month%3D02/day%3D22/server%3Dbangkok/2020-02-22-13-00-07-386017311-97cb2717-squid.log.gz (The write operation timed out)
WARNING: Waiting 6 sec...
Fatal Python error: GC object already tracked

Current thread 0x00007fcf1dc31700 (most recent call first):
```

This attempt segfaulted after giving up!
```
WARNING: Upload failed: /year%3D2020/month%3D02/day%3D22/server%3Dbangkok/2020-02-22-13-00-07-386017311-97cb2717-squid.log.gz (400 (RequestTimeout): Your socket connection to the server was not read from or written to within the timeout period. Idle connections will be closed.)
WARNING: Waiting 3 sec...
WARNING: Retrying failed request: /year%3D2020/month%3D02/day%3D22/server%3Dbangkok/2020-02-22-13-00-07-386017311-97cb2717-squid.log.gz ([Errno 0] Error) 
WARNING: Waiting 6 sec...
WARNING: Upload failed: /year%3D2020/month%3D02/day%3D22/server%3Dbangkok/2020-02-22-13-00-07-386017311-97cb2717-squid.log.gz (400 (RequestTimeout): Your socket connection to the server was not read from or wr
itten to within the timeout period. Idle connections will be closed.)
WARNING: Waiting 9 sec...
WARNING: Upload failed: /year%3D2020/month%3D02/day%3D22/server%3Dbangkok/2020-02-22-13-00-07-386017311-97cb2717-squid.log.gz (400 (RequestTimeout): Your socket connection to the server was not read from or written to within the timeout period. Idle connections will be closed.)
WARNING: Waiting 12 sec...
WARNING: Upload failed: /year%3D2020/month%3D02/day%3D22/server%3Dbangkok/2020-02-22-13-00-07-386017311-97cb2717-squid.log.gz (400 (RequestTimeout): Your socket connection to the server was not read from or written to within the timeout period. Idle connections will be closed.)
WARNING: Waiting 15 sec...
ERROR: Upload of '/var/log/squid/s3/2020-02-22-13-00-07-386017311-97cb2717-squid.log.gz' failed too many times (Last reason: Upload failed for: /year%3D2020/month%3D02/day%3D22/server%3Dbangkok/2020-02-22-13-00-07-386017311-97cb2717-squid.log.gz)
```

Here's another segfault:
```
WARNING: Upload failed: /year%3D2020/month%3D02/day%3D22/server%3Dbangkok/2020-02-22-13-00-07-386017311-97cb2717-squid.log.gz (400 (RequestTimeout): Your socket connection to the server was not read from or written to within the timeout period. Idle connections will be closed.)
WARNING: Waiting 3 sec...
WARNING: Upload failed: /year%3D2020/month%3D02/day%3D22/server%3Dbangkok/2020-02-22-13-00-07-386017311-97cb2717-squid.log.gz (400 (RequestTimeout): Your socket connection to the server was not read from or written to within the timeout period. Idle connections will be closed.)
WARNING: Waiting 6 sec...
WARNING: Upload failed: /year%3D2020/month%3D02/day%3D22/server%3Dbangkok/2020-02-22-13-00-07-386017311-97cb2717-squid.log.gz (The write operation timed out)
WARNING: Waiting 9 sec...
```

**Comment by fviard** [comment]
Which version of Python are you using?

**Comment by bigwillystyle42** [comment]
We are using python 3.5.3-1 from the Debian 9 repositories.

I can provide information, but I'm no longer able to help with any testing. We switched to using the standard awscli package a week ago.

**Comment by fviard** [comment]
That is curious, but should not come directly from s3cmd code but from a c-extension module of python.
Hum that is sad, something that could have help would have been to run s3cmd with the "-v -d" flag on one of the affected instance to see in the detailed log what operation was causing that.

**Comment by fviard** [comment]
I think that this issue might now be fixed with #1114
